x/merkledb/trie.go

```diff
@@ -34,46 +66,225 @@ type ReadOnlyTrie interface {
 	// database.ErrNotFound if the key is not present
 	GetValues(ctx context.Context, keys [][]byte) ([][]byte, []error)
 
-	// get the value associated with the key in path form
-	// database.ErrNotFound if the key is not present
-	getValue(key Key) ([]byte, error)
-
-	// get an editable copy of the node with the given key path
-	// hasValue indicates which db to look in (value or intermediate)
-	getEditableNode(key Key, hasValue bool) (*node, error)
-
 	// GetRangeProof returns a proof of up to [maxLength] key-value pairs with
 	// keys in range [start, end].
 	// If [start] is Nothing, there's no lower bound on the range.
 	// If [end] is Nothing, there's no upper bound on the range.
 	GetRangeProof(ctx context.Context, start maybe.Maybe[[]byte], end maybe.Maybe[[]byte], maxLength int) (*RangeProof, error)
 
-	database.Iteratee
-}
-
-type ViewChanges struct {
-	BatchOps []database.BatchOp
-	MapOps   map[string]maybe.Maybe[[]byte]
-	// ConsumeBytes when set to true will skip copying of bytes and assume
-	// ownership of the provided bytes.
-	ConsumeBytes bool
-}
-
-type Trie interface {
-	ReadOnlyTrie
-
 	// NewView returns a new view on top of this Trie where the passed changes
 	// have been applied.
 	NewView(
 		ctx context.Context,
 		changes ViewChanges,
-	) (TrieView, error)
+	) (View, error)
 }
 
-type TrieView interface {
+type View interface {
 	Trie
 
 	// CommitToDB writes the changes in this view to the database.
 	// Takes the DB commit lock.
 	CommitToDB(ctx context.Context) error
 }
+
+// isSentinelNodeTheRoot returns true if the sentinel node of the passed in trie has a value and or multiple child nodes
+// When this is true, the root of the trie is the sentinel node
+// When this is false, the root of the trie is the sentinel node's single child
+func isSentinelNodeTheRoot[T Trie](t T) bool {
```

Comment:
They aren't strictly required, but theoretically the generics allow for static dispatch instead of dynamic dispatch of the functions and I'm trying to squeeze any perf gains I can.